File: Signiture-Backend/kotlin/study/src/main/java/clazz/Exam.java
package clazz;

public class Exam {

    interface Bark {
        void bark();
    }

    abstract class Animal implements Bark {

        private final String name;

        public Animal(String name) {
            this.name = name;
        }

        public void eat() {
            System.out.println(name + " 식사 시작 합니다.");
        }

        public String getName() {
            return name;
        }
    }

    class Dog extends Animal {

        private int age;

        public Dog() {
            super("Dog");
        }

        public Dog(String name) {
            super(name);
        }

        @Override
        public void bark() {
            System.out.println(this.getName() + ": 멍멍!!");
        }

        public int getAge() {
            return age;
        }

        public void setAge(int age) {
            this.age = age;
        }
    }

    public Exam() {
        var dog = new Dog("바둑이");
        dog.eat();
        dog.bark();

        dog.setAge(10);
        System.out.println("age: " + dog.getAge());

        var user = new User();
        user.setName("홍길동");
        user.setEmail("email@gmail.com");
        user.setAge(20);
        System.out.println(user);

        // 생성자 매개변수의 순서를 준수 해야 한다.
        new User("홍길동", "email@gmail.com", 20);
    }

    public static void main(String[] args) {
        new Exam();
    }
}

File: Signiture-Backend/kotlin/study/src/main/kotlin/clazz/Exam.kt
package pcy.study.clazz

interface Bark {
    fun bark()
}

abstract class Animal(
    private val name: String? = "Dog"
) : Bark {

    // body
    fun eat() {
        println("$name 식사 시작 합니다.")
    }
}

class Dog(
    val name: String? = null,
) : Animal(name) {

    var age: Int? = null

    override fun bark() {
        println("$name: 멍멍!!")
    }
}

fun main() {
    val dog = Dog("바둑이")
    dog.eat()
    dog.bark()

    dog.age = 10;
    println("age: ${dog.age}")

    val user = User()
    user.name = "홍길동"
    user.email = "email@gmail.com"
    user.age = 20
    println(user)

    val copyUser = user.copy(name = "복제 사용자")
    println(copyUser)

    // Named Argument를 사용하면 매개변수를 직접 지정할 수 있다.
    User(
        age = 20,
        name = "홍길동",
        email = "email@gmail.com",  // 마지막 쉼표(,)에 대해 컴파일 에러가 발생하지 않는다.
    )
}
